[Flang][OpenMP] Additional global address space modifications for device

agozillon opened the pull request:
A prior PR added a portion of the global address space modifications required for declare target to, this PR seeks to add
 a small amount more leftover from that PR.
   
The intent is to allow for more correct IR that the backends (in particular AMDGPU) can treat more aptly for optimisations and code correctness
    
1/3 required PRs to enable declare target to mapping, should look at PR 3/3 to check for full green passes (this one will fail a number due to some dependencies).
    
Co-authored-by: Raghu Maddhipatla raghu.maddhipatla@amd.com

agozillon commented:
Small ping for some attention on this PR if at all possible please :-)

abidh commented on flang/lib/Optimizer/CodeGen/FIROpPatterns.cpp:
```diff
@@ -346,7 +346,10 @@ unsigned ConvertFIRToLLVMPattern::getAllocaAddressSpace(
+  auto module = mlir::isa<mlir::ModuleOp>(parentOp)
+                    ? mlir::cast<mlir::ModuleOp>(parentOp)
+                    : parentOp->getParentOfType<mlir::ModuleOp>();
+  if (module)
```
This change and the one below is not exactly required for this PR. I guess they are small enough to include in this but could go in a separate PR.

abidh reviewed (approved):
Thanks for working on this. I recently hit this issue (address space of global) as well while working on the debug for declare target variables. Looks ok to me mostly apart from a few nits.

agozillon reviewed (commented):
@abidh Addressed your comments as best I could, hopefully they're reasonable changes! :-) Unfortunately, I've also done a rebase which breaks the PR as it brings in some incompatible changes that'll need fixed in a seperate PR as it's related to extending the getDataLayout utilities to support the new GPU module that's occasionally in use. But it doesn't make sense to add it in this PR I think, so I'll do so in a seperate one I'll open in the near future (hopefully today) :-) And thank you very much for the review! Took me a bit to get around to addressing them as I'd hoped the other segments of the stack would have received some reviews by now so I could do it all at once, alas it was not to be!

abidh commented:
> @abidh Addressed your comments as best I could, hopefully they're reasonable changes! :-) Unfortunately, I've also done a rebase which breaks the PR as it brings in some incompatible changes that'll need fixed in a seperate PR as it's related to extending the getDataLayout utilities to support the new GPU module that's occasionally in use. But it doesn't make sense to add it in this PR I think, so I'll do so in a seperate one I'll open in the near future (hopefully today) :-) And thank you very much for the review! Took me a bit to get around to addressing them as I'd hoped the other segments of the stack would have received some reviews by now so I could do it all at once, alas it was not to be!

Thanks. I skimmed through your changes and they look ok to me.

skatrak commented on flang/lib/Optimizer/CodeGen/CodeGen.cpp:
```diff
@@ -1350,14 +1399,34 @@ struct EmboxCommonConversion : public fir::FIROpConversion<OP> {
+    // like it was intended to).
+    std::optional<mlir::DataLayout> dataLayout =
+        fir::support::getOrSetMLIRDataLayout(
+            mod, /*allowDefaultLayout*/ mlir::isa<mlir::gpu::GPUModuleOp>(mod)
+                     ? false
+                     : true);
```
```suggestion
            mod, /*allowDefaultLayout=*/!mlir::isa<mlir::gpu::GPUModuleOp>(mod));
```

skatrak commented on flang/lib/Optimizer/CodeGen/CodeGen.cpp:
```diff
@@ -1350,14 +1399,34 @@ struct EmboxCommonConversion : public fir::FIROpConversion<OP> {
             : fir::NameUniquer::getTypeDescriptorName(recType.getName());
     mlir::Type llvmPtrTy = ::getLlvmPtrType(mod.getContext());
+
+    // We allow the module to be set to a default layout if it's a regular module
+    // however, we prevent this if it's a GPU module, as the datalayout in these
+    // cases will currently be the union of the GPU Module and the parent builtin
+    // module, with the GPU module overriding the parent where there are duplicates.
+    // However, if we force the default layout onto a GPU module, with no datalayout
+    // it'll result in issues as the infrastructure does not support the union of
+    // two layouts with builtin data layout entries currently (and it doesn't look
+    // like it was intended to).
```
```suggestion
    // We allow the module to be set to a default layout if it's a regular module.
    // However, we prevent this if it's a GPU module, as the datalayout in these
    // cases will currently be the union of the GPU Module and the parent builtin
    // module, with the GPU module overriding the parent where there are duplicates.
    // However, if we force the default layout onto a GPU module, with no datalayout
    // it'll result in issues as the infrastructure does not support the union of
    // two layouts with builtin data layout entries.
```

skatrak commented on flang/lib/Optimizer/CodeGen/CodeGen.cpp:
```diff
@@ -134,16 +134,65 @@ addLLVMOpBundleAttrs(mlir::ConversionPatternRewriter &rewriter,
+                                      std::uint64_t programAS,
+                                      llvm::StringRef symName, mlir::Type type,
+                                      mlir::Operation *replaceOp) {
+  if (mlir::isa<mlir::LLVM::LLVMPointerType>(type)) {
+    if (globalAS != programAS) {
+      auto llvmAddrOp = rewriter.create<mlir::LLVM::AddressOfOp>(
+          loc, getLlvmPtrType(rewriter.getContext(), globalAS), symName);
+      return rewriter.replaceOpWithNewOp<mlir::LLVM::AddrSpaceCastOp>(
+          replaceOp, ::getLlvmPtrType(rewriter.getContext(), programAS),
+          llvmAddrOp);
+    }
+    return rewriter.replaceOpWithNewOp<mlir::LLVM::AddressOfOp>(
+        replaceOp, getLlvmPtrType(rewriter.getContext(), globalAS), symName);
+  }
+  return rewriter.replaceOpWithNewOp<mlir::LLVM::AddressOfOp>(replaceOp, type,
+                                                              symName);
```
Nit: Feel free to disagree, but it seems like the logic here could be simplified a bit with something like this:
```suggestion
  bool isPointerType = mlir::isa<mlir::LLVM::LLVMPointerType>(type);
  if (isPointerType)
    type = getLlvmPtrType(rewriter.getContext(), globalAS)

  auto llvmAddrOp = rewriter.create<mlir::LLVM::AddressOfOp>(
      loc, type, symName);

  if (isPointerType && globalAS != programAS)
    return rewriter.replaceOpWithNewOp<mlir::LLVM::AddrSpaceCastOp>(
      replaceOp, getLlvmPtrType(rewriter.getContext(), programAS),
      llvmAddrOp);

  return rewriter.replaceOp(replaceOp, llvmAddrOp);
```

skatrak commented on flang/lib/Optimizer/CodeGen/CodeGen.cpp:
```diff
@@ -134,16 +134,65 @@ addLLVMOpBundleAttrs(mlir::ConversionPatternRewriter &rewriter,
+                                 std::uint64_t programAS,
+                                 llvm::StringRef symName, mlir::Type type) {
+  if (mlir::isa<mlir::LLVM::LLVMPointerType>(type)) {
+    if (globalAS != programAS) {
+      auto llvmAddrOp = rewriter.create<mlir::LLVM::AddressOfOp>(
+          loc, getLlvmPtrType(rewriter.getContext(), globalAS), symName);
+      return rewriter.create<mlir::LLVM::AddrSpaceCastOp>(
+          loc, getLlvmPtrType(rewriter.getContext(), programAS), llvmAddrOp);
+    }
+    return rewriter.create<mlir::LLVM::AddressOfOp>(
+        loc, getLlvmPtrType(rewriter.getContext(), globalAS), symName);
```
```suggestion
    auto llvmAddrOp = rewriter.create<mlir::LLVM::AddressOfOp>(
        loc, getLlvmPtrType(rewriter.getContext(), globalAS), symName);
    if (globalAS != programAS)
      return rewriter.create<mlir::LLVM::AddrSpaceCastOp>(
          loc, getLlvmPtrType(rewriter.getContext(), programAS), llvmAddrOp);
    return llvmAddrOp;
```

skatrak reviewed (approved):
I have some minor nits and suggestions to hopefully improve code readability, but LGTM.

agozillon committed:
[Flang][OpenMP] Additional global address space modifications for device

A prior PR added a portion of the global address space modifications required for declare target to, this PR seeks to add a small amount more leftover from that PR.

The intent is to allow for more correct IR that the backends (in particular AMDGPU) can treat more aptly for optimisations and code correctness

1/3 required PRs to enable declare target to mapping, should look at PR 3/3 to check for full green passes (this one will fail a number due to some dependencies).

Co-authored-by: Raghu Maddhipatla raghu.maddhipatla@amd.com

skatrak reviewed (approved):
LGTM!

jsjodin reviewed (approved):
LGTM

agozillon commented:
As this PR has the pre-requisite approvals and isn't dependent on any other PR in the stack, is passing the tests (currently) and it turns out there's a broken check-offload test as we're missing a few bits from here I'm going to rebase and land it :-)

agozillon committed:
Merge branch 'main' into users/agozillo/declare-target-to-1
